pkg/controllers/jobset_controller.go

```diff
@@ -286,7 +234,175 @@ func (r *JobSetReconciler) createHeadlessSvcIfNotExist(ctx context.Context, js *
+		// To reach this point a job must have failed, and TerminationPolicyTargetAny applies to any job
+		// in the JobSet, so we can skip directly to executing the restart policy.
+		return r.failureRestartPolicy(ctx, js, ownedJobs)
+	default:
+		return fmt.Errorf("invalid termination policy target %s", js.Spec.FailurePolicy.Target)
+	}
+}
+
+func (r *JobSetReconciler) failureRestartPolicy(ctx context.Context, js *jobset.JobSet, ownedJobs *childJobs) error {
+	switch js.Spec.FailurePolicy.RestartPolicy {
+	case jobset.RestartPolicyRecreateAll:
+		return r.restartPolicyRecreate(ctx, js, ownedJobs)
+	case jobset.RestartPolicyNone:
+		return r.updateStatusCondition(ctx, js, metav1.Condition{
+			Type:               string(jobset.JobSetFailed),
+			Status:             metav1.ConditionStatus(corev1.ConditionTrue),
+			LastTransitionTime: metav1.Now(),
+			Reason:             "FailedJobs",
+			Message:            "jobset failed due to one or more job failures",
+		})
+	default:
+		return fmt.Errorf("invalid restart policy: %s", js.Spec.FailurePolicy.RestartPolicy)
+	}
+}
+
+func (r *JobSetReconciler) restartPolicyRecreate(ctx context.Context, js *jobset.JobSet, ownedJobs *childJobs) error {
+	log := ctrl.LoggerFrom(ctx).WithValues("jobset", klog.KObj(js))
+	ctx = ctrl.LoggerInto(ctx, log)
+
+	// If JobSet has reached max number of restarts, mark it as failed and return.
+	if js.Status.Restarts == js.Spec.FailurePolicy.MaxRestarts {
+		return r.updateStatusCondition(ctx, js, metav1.Condition{
+			Type:               string(jobset.JobSetFailed),
+			Status:             metav1.ConditionStatus(corev1.ConditionTrue),
+			LastTransitionTime: metav1.Now(),
+			Reason:             "ReachedMaxRestarts",
+			Message:            "jobset failed due to reaching max number of restarts",
+		})
+	}
+
+	// Increment JobSet restarts.
+	js.Status.Restarts += 1
+	if err := r.Status().Update(ctx, js); err != nil {
+		return fmt.Errorf("updating jobset status to increment restarts: %v", err)
+	}
+
+	// Delete jobs marked for deletion. Deletions will trigger reconciliation events, which will result in
+	// the jobs being recreated.
+	return r.deleteJobs(ctx, js, ownedJobs)
+}
+
+// deletes all jobs in parallel.
+func (r *JobSetReconciler) deleteJobs(ctx context.Context, js *jobset.JobSet, ownedJobs *childJobs) error {
+	log := ctrl.LoggerFrom(ctx).WithValues("jobset", klog.KObj(js))
+	ctx = ctrl.LoggerInto(ctx, log)
+
+	var wg sync.WaitGroup
+	var finalErr error
+	delete := ownedJobs.failed
+	for _, job := range ownedJobs.active {
+		// Jobs with jobset.sigs.k8s.io/restart-attempt < jobset.status.restarts are marked for
+		// deletion, as they were part of the previous JobSet lifecycle iteration.
+		jobRestarts, err := strconv.Atoi(job.Labels[jobset.RestartsLabel])
+		if err != nil {
+			return err
+		}
+		if jobRestarts < js.Status.Restarts {
+			delete = append(delete, job)
+		}
+	}
+
+	// Delete all jobs in parallel.
+	for _, job := range delete {
```

Comment:
we need to limit the number of go routines we create, perhaps add a comment to address this in a followup PR. we probably need a common function to delete/create jobs in parallel.